app/javascript/gameStatesFactory.js

```diff
@@ -0,0 +1,304 @@
+    }
+    if(gameStates.selfMoves[gameId].length === 0) {
+      throw 'This game has no self moves yet';
+    }
+
+    for(let currentMove of gameStates.selfMoves[gameId]) {
+      let currentMoveNumber = gameStates.getMoveNumberFromState(currentMove.newState);
+
+      if(currentMoveNumber === moveNumber) {
+        return currentMove;
+      }
+    }
+  };
+
+  gameStates.getPreviousOpponentMove = function(gameId) {
+    if(typeof(gameStates.opponentMoves[gameId]) === 'undefined') {
+      throw 'This game has no moves yet';
+    }
+    if(gameStates.opponentMoves[gameId].length === 0) {
+      throw 'This game has no opponent moves yet';
+    }
+    if(gameStates.opponentMoves[gameId].length < 2) {
+      throw 'This game has no previous opponent move';
+    }
+  };
+  gameStates.getLastOpponentMove = function(gameId) {
+    if(typeof(gameStates.opponentMoves[gameId]) === 'undefined') {
+      throw 'This game has no moves yet';
+    }
+    if(gameStates.opponentMoves[gameId].length === 0) {
+      throw 'This game has no opponent moves yet';
+    }
+
+    return gameStates.opponentMoves[gameId][gameStates.opponentMoves[gameId].length - 1];
+  };
+
+  /**
+   *
+   * @param gameId
+   * @return boolean
+     */
+  gameStates.isBlockchainStateNewer = function(gameId) {
+    let blockchainState = Chess.getCurrentGameState(gameId);
+
+    if(blockchainState) {
+      if(typeof(gameStates.lastMoveNumber[gameId]) === 'undefined') {
+        // No state in the local states, so return true
+        return true;
+      }
+
+      let blockchainMoveNumber = gameStates.getMoveNumberFromState(blockchainState);
+
+      return blockchainMoveNumber > gameStates.lastMoveNumber[gameId];
+    }
+    // No state in the blockchain, so return false
+    return false;
+  };
+
+  /**
+   * Get an array with all parameters needed to send them to the blockchain.
+   * Throws if the last move was not done by oneself.
+   * @param gameId
+   * @return [oldOpponentState, oldOpponentStateSignature, moveSelfFrom, moveSelfTo]
+     */
+  gameStates.getSelfMovePackage = function(gameId) {
+    if(typeof(gameStates.selfMoves[gameId]) === 'undefined') {
+      throw 'This game has no moves yet';
+    }
+    if(gameStates.selfMoves[gameId].length === 0) {
+      throw 'This game has no self moves yet';
+    }
+
+    let lastSelfMove = gameStates.selfMoves[gameId][gameStates.selfMoves[gameId].length - 1];
+    let lastSelfMoveNumber = gameStates.getMoveNumberFromState(lastSelfMove.newState);
+
+    if(lastSelfMoveNumber !== gameStates.lastMoveNumber[gameId]) {
+      throw 'The self move was not the last move';
+    }
+
+    // There was only one move, and it was the self move => this was our first move as white
+    if(gameStates.lastMoveNumber[gameId] === 1) {
+      return [null, null, lastSelfMove.moveFrom, lastSelfMove.moveTo];
+    }
+
+    let lastOpponentMove = gameStates.opponentMoves[gameId][
+      gameStates.opponentMoves[gameId].length - 1
```

Comment:
This will fail if the array is empty (or there is no such array -> don't know if this can be the case :D).  
You should also check if the state and the move have successive move counters.